### PR TITLE
Quark Server and Thin Jdbc Client

### PR DESCRIPTION
**Pull request by adeshr**

### CODE REVIEW

**Comment by adeshr** [comment]
@vrajat  @amoghmargoor 
1) Need input about how to dynamically pass args to server
   - EndToEndTest will change according to arguments passed to the server
2) Python script to start the server is not yet working (some problem with daemon I think), working on it.

File: server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        + "     \"url\":\"jdbc:mysql://localhost.localdomain:3306/nezha_rstore\","
+        + "     \"username\":\"root\","
+        + "     \"password\":\"\","
+        + "     \"encrypt_key\":\"easy\""
+        + "   }";
+    props.put("dbCredentials", jsonString);
+    props.put("schemaFactory", "com.qubole.quark.catalog.db.SchemaFactory");
+
+    try {
+      final String url;
```
Comment (vrajat) [inline server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java]: Is `url` used ?

**Comment by vrajat** [comment]
Can you explain what the python files do ?

**Comment by vrajat** [comment]
The simplest option is to start the java process with path to a json file as a command line argument. Similar to Nezha or Scheduler within QDS

**Comment by vrajat** [comment]
I am interested in your test. Do you start a server or is the server running within the same JVM ?

**Comment by adeshr** [comment]
Python Files:
quark_utils.py finds the location of server/client jar, which will be used by server.py
server.py actually starts/stops the server.
daemon.py is used by server.py.

Test:
server is running within same JVM, so that the H2 db is accessible by both client/server.

**Comment by vrajat** [comment]
I dont think we need `quark_utils.py`

File: server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      ClassNotFoundException {
+    setupTables(h2Url, "tpcds.sql");
+    setupTables(cubeUrl, "tpcds_cubes.sql");
+    setupTables(viewUrl, "tpcds_views.sql");
+  }
+
+  public static void setupTables(String dbUrl, String filename)
+      throws ClassNotFoundException, SQLException, IOException, URISyntaxException {
+
+    new Thread(new Main()).start();
```
Comment (vrajat) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: Oh - This is where you start the server

Comment (vrajat) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: Ah! So even in this test it is reading the mysql store.

Comment (adeshr) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: Yes, I have to change it to H2 Db, was waiting for how/which arguments to pass to server.

**Comment by vrajat** [comment]
Can you pass in args from the test ? The arg will be path to a json file in resources dir.

**Comment by adeshr** [comment]
sure, will try that.

File: server/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+
+  <properties>
+    <!-- check if there is a better way -->
+    <project.rootdir>${basedir}/..</project.rootdir>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.calcite</groupId>
```
Comment (amoghmargoor) [inline server/pom.xml]: @vrajat Should we just use our calcite ?

File: server/src/main/java/com/qubole/quark/serverclient/server/Main.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
+/*
```
Comment (amoghmargoor) [inline server/src/main/java/com/qubole/quark/serverclient/server/Main.java]: Be careful about copyright note ...if you have copied significant parts from any project ..make sure you retain the copyright note too. Don't alter that.

File: server/src/main/java/com/qubole/quark/serverclient/server/Main.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  public void awaitRunning(long timeout, TimeUnit unit) throws InterruptedException {
+    runningLatch.await(timeout, unit);
+  }
+
+  public void run(String[] args) throws Exception {
+    logJVMInfo();
+    try {
+
+      Class<? extends Meta.Factory> factoryClass = QuarkMetaFactoryImpl.class;
+      int port = 8765;
```
Comment (amoghmargoor) [inline server/src/main/java/com/qubole/quark/serverclient/server/Main.java]: Make this configurable (whenever you make changes to pass relevant arguments to server)

File: server/pom.xml (hunk 3 above)
Comment (vrajat) [inline server/pom.xml]: yes

File: client/pom.xml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+-->
+
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>quark-parent</artifactId>
+    <groupId>com.qubole</groupId>
+    <version>4.3.1</version>
```
Comment (vrajat) [inline client/pom.xml]: Version should be `4.4.0-SNAPSHOT`. Please change it everywhere. You may have to fetch upstream

File: server/pom.xml (hunk 3 above)
Comment (vrajat) [inline server/pom.xml]: @adeshr Can you check with Quark's version of Calcite ?

File: server/src/main/java/com/qubole/quark/serverclient/server/Main.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A query server for Phoenix over Calcite's Avatica.
```
Comment (vrajat) [inline server/src/main/java/com/qubole/quark/serverclient/server/Main.java]: Fix Comments

File: server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  public QuarkMetaFactoryImpl() {
+    super();
+  }
+
+  @Override
+  public Meta create(List<String> args) {
+    Properties props = new Properties();
+    String url = "jdbc:quark:";
+    try {
+      if (args.size() == 1 || args.size() == 2) {
```
Comment (vrajat) [inline server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java]: Why will arg size be equal to 2 ? I see only one call `args.get(0)`

Comment (adeshr) [inline server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java]: for port number, args.get(1) is in Main.java

File: server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  public static final String h2Url = "jdbc:h2:mem:TpcdsTest;DB_CLOSE_DELAY=-1";
+  public static final String cubeUrl = "jdbc:h2:mem:TpcdsCubes;DB_CLOSE_DELAY=-1";
+  public static final String viewUrl = "jdbc:h2:mem:TpcdsViews;DB_CLOSE_DELAY=-1";
+
+  @BeforeClass
+  public static void setUp() throws SQLException, IOException, URISyntaxException,
+      ClassNotFoundException {
+    String[] args = new String [1];
+    args[0] = "jsonCatalog.json";
+    new Thread(new Main(args)).start();
```
Comment (vrajat) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: Will need one more test with a DB catalog. DB can be H2

**Comment by vrajat** [comment]
I have added some comments. All incremental stuff. I'll give it a spin.

File: server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java (hunk 9 above)
Comment (adeshr) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: but that will need 12 migration files in resources, we already have two copies one jdbc module and other in catalog module.

File: server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java (hunk 8 above)
Comment (vrajat) [inline server/src/main/java/com/qubole/quark/serverclient/server/QuarkMetaFactoryImpl.java]: I suggest putting that into the config file.

File: server/pom.xml (hunk 3 above)
Comment (vrajat) [inline server/pom.xml]: @adeshr Ping

**Comment by vrajat** [comment]
@adeshr Can you provide instructions to start the server ?

File: server/pom.xml (hunk 3 above)
Comment (adeshr) [inline server/pom.xml]: I made the change in calcite-avatica-server, forgot to do it here. will change it

**Comment by adeshr** [comment]
sudo ./bin/server.py start <filename>
sudo ./bin/server.py stop

File: server/pom.xml [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+      <artifactId>calcite-avatica</artifactId>
+      <version>${calcite.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.calcite</groupId>
+      <artifactId>calcite-avatica-server</artifactId>
+      <version>${calcite.version}</version>
+    </dependency>
+    <dependency>
```
Comment (vrajat) [inline server/pom.xml]: Please use logback instead. We use logback in other modules.

```
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-core</artifactId>
            <version>${logback-version}</version>
        </dependency>
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>${logback-version}</version>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>jcl-over-slf4j</artifactId>
            <version>1.7.13</version>
        </dependency>
```

File: bin/server.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+if java_home:
+    java = os.path.join(java_home, 'bin', 'java')
+else:
+    java = 'java'
+
+# The command is run through subprocess so environment variables are automatically inherited
+java_cmd = '%(java)s -cp ' + \
+           quark_server_jar + \
+           " -Dproc_quarkserver" + \
+           " -Dpsql.root.logger=%(root_logger)s" + \
```
Comment (vrajat) [inline bin/server.py]: The logging parameters will be invalid once you move to logback. So this has to be changed.

File: client/pom.xml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>quark-parent</artifactId>
+    <groupId>com.qubole</groupId>
+    <version>4.4.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>quark-server-client</artifactId>
```
Comment (vrajat) [inline client/pom.xml]: Please rename the jar to `quark-client`

**Comment by vrajat** [comment]
The package and jars names have to be cleaned up. @amoghmargoor please comment:

1. Server
   Jar: `quark-server-<version>.jar`
   Package: `com.qubole.quark.server`

2. Client
    Jar: `quark-thin-jdbc-<version>.jar`
    Package: `com.qubole.quark.jdbc.thin`

Actually, I prefer the thin JDBC jar to be the preferred one. So I prefer the thin client to be: 
    Jar: `quark-jdbc-<version>.jar`
   Package: `com.qubole.quark.jdbc`

And the fat jar to be:
    Jar: `quark-jdbc-with-server.jar`
   Package: `com.qubole.quark.jdbcserver`

So using the fat jar is _icky_ which is good.

File: server/src/test/java/com/qubole/quark/serverclient/client/DbEndToEndTest.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    viewUrl = "jdbc:h2:mem:DbServerTpcdsViews;DB_CLOSE_DELAY=-1";
+  }
+
+  @BeforeClass
+  public static void setUp() throws SQLException, IOException, URISyntaxException,
+      ClassNotFoundException {
+
+    new Thread(main).start();
+
+    Flyway flyway = new Flyway();
```
Comment (vrajat) [inline server/src/test/java/com/qubole/quark/serverclient/client/DbEndToEndTest.java]: Hmm - There is something fishy going on. When I ran the server, I noticed that the migrations did not run. This seems to be same symptom. Ideally server should run it on start up.

File: server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  public void testClient() throws SQLException, ClassNotFoundException, InterruptedException {
+    Class.forName("com.qubole.quark.jdbc.QuarkDriver");
+    QuarkDriver d = new QuarkDriver();
+    Class.forName("org.h2.Driver");
+    Connection connection = null;
+
+    // Due to threading, server might not be up.
+    for (int i=0; i<2; i++) {
+       try {
+         connection = d.connect(ThinClientUtil.getConnectionUrl("0.0.0.0", 8765), new Properties());
```
Comment (vrajat) [inline server/src/test/java/com/qubole/quark/serverclient/client/EndToEndTest.java]: The test is fishy. You should be using DriverManager to connect. So this part has to be parameterized depending on whether you are running `DbEndToEnd` or `JsonEndToEnd`. 
Example: https://github.com/qubole/quark/blob/master/jdbc/src/test/java/com/qubole/quark/jdbc/test/DbSelectTest.java#L67

**Comment by amoghmargoor** [comment]
@vrajat +1 for names suggested. I have one clarification though... isn't quark-server a jdbc-server and not the fat jar ?

**Comment by vrajat** [comment]
Thanks @amoghmargoor 

@adeshr I have updated my previous comment with the names we decided after an offline discussion.

**Comment by vrajat** [comment]
@adeshr Regarding tests. Please add a suite that does the following:

1. Start server in `setupClass`
2. Test database is empty and there are no tables. Check H2 docs to get a list of tables.
3. Connect and make sure migrations have run by checking the latest migration in `versions` table.

**Comment by vrajat** [comment]
FYI
I am not able to use the thin client with SQLLine.

```
sqlline> !connect jdbc:quark:thin:url=http://localhost:8765;serialization=PROTOBUF com.qubole.quark.serverclient.client.QuarkDriver
scan complete in 7ms
scan complete in 242ms
No known driver to handle "jdbc:quark:thin:url=http://localhost:8765;serialization=PROTOBUF"
sqlline>
```

I'll wait for you to fix your tests before triaging it.

**Comment by adeshr** [comment]
I am able to connect to sqlline, but have to add calcite-avatica.jar in classpath.
I think, previously, we used to connect through fatjdbc.jar which had all the required jar.

**Comment by vrajat** [comment]
Can you shade the avatica jar ?

**Comment by adeshr** [comment]
done.

**Comment by vrajat** [comment]
You are right. I think you should go ahead and merge it without fixing the test. I am tracking the fixes required in #40 & #41.

One last request. Can you rename directories:
`jdbc` -> `fatjdbc`
`client` -> `jdbc`

**Comment by vrajat** [comment]
If its OK with @amoghmargoor please prepare the branch for a merge. Please use the following git commit message (we need a process where the author suggests the commit message going forward)

```
new:usr: #20 Quark Server and Thin JDBC Client

Quark is available as a JDBC Server and a thin JDBC client. Calcite's Avatica
framework is used to implement the server and client. The client is very small in 
size and can be easily distributed. The server accepts a JSON configuration and
connects to a DB Catalog. JSON Catalog is not supported with this setup.

JDBC directories and packages have been renamed to fatjdbc.
The new client is the official JDBC driver going forward. It uses the jdbc namespace.
```

File: plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java [hunk 15]
```diff
@@ -33,7 +33,7 @@ public void testGenericDb() throws QuarkException {
   public void testValidProps() throws QuarkException {
     Map<String, Object> props = new HashMap<>();
     props.put("type", "mysql");
-    props.put("url", "jdbc:quark:test");
+    props.put("url", "fatjdbc:quark:test");
```
Comment (amoghmargoor) [inline plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java]: Have you changed the connect string for fat jar from jdbc:quark to fatjdbc:quark ?

Comment (vrajat) [inline plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java]: Oh yea - that doesn't look right. Should it be 'jdbc:quark:fat:' ?

Comment (adeshr) [inline plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java]: must have missed it during module rename, fixing it.

**Comment by amoghmargoor** [comment]
LGTM apart from one minor nit. Ship it from my side.

File: plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java (hunk 15 above)
Comment (adeshr) [inline plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java]: Also, connect string is "jdbc:quark:" for fatjdbc, and "jdbc:quark:thin:" for jdbc. I didn't make any changes in fatjdbc connect string.

Comment (vrajat) [inline plugins/src/test/java/com/qubole/quark/plugins/jdbc/JdbcFactoryValidationTest.java]: Change that to what I suggested. Hopefully it's a global search replace

**Comment by adeshr** [comment]
Sure @amoghmargoor .

**Pull request merged**